### PR TITLE
Fix performance issues

### PR DESCRIPTION
**Pull request by LahiruLS**
Improve codebase and the DB metadata to popup feature performance

### CODE REVIEW

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java [hunk 1]
```diff
@@ -34,7 +34,7 @@
 import java.util.List;
 import java.util.Map;
 
-import static org.wso2.carbon.identity.organization.mgt.core.constant.OrganizationMgtConstants.PATCH_PATH_ORG_PARENT_ID;
+import static org.wso2.carbon.identity.organization.mgt.core.constant.OrganizationMgtConstants.*;
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java:37]: No wild card imports please

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java [hunk 2]
```diff
@@ -236,6 +272,8 @@ private void clearOrganizationCache(int tenantId, String organizationId) throws
             }
             OrganizationCacheKey cacheKey = new OrganizationCacheKey(organizationId);
             organizationCache.clearCacheEntry(cacheKey);
+            organizationCache.clearCacheEntry(
+                    new OrganizationCacheKey(String.format(ALL_LEVEL_CHILD_ORG_LIST_CACHE_KEY_FORMAT, organizationId)));
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java:276]: Will this cache get invalidated/updated upon org add/deletion?

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java [hunk 3]
```diff
@@ -132,6 +133,18 @@ Organization getOrganization(int tenantId, String organizationId, String userId,
      */
     List<String> getChildOrganizationIds(String organizationId, String userId) throws OrganizationManagementException;
 
+    /**
+     * Fetch All the child organization IDs for a specified parent organization.
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java:137]: 'all'

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java [hunk 4]
```diff
@@ -132,6 +133,18 @@ Organization getOrganization(int tenantId, String organizationId, String userId,
...
+     *
+     * @param organizationId
+     * @return
+     * @throws OrganizationManagementException
+     */
+    default List<String> getAllOfChildOrganizationIds(String organizationId) throws OrganizationManagementException {
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java:143]: How about 'getAllCascadingChildOrganizationIds' ?

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java [hunk 5]
```diff
@@ -132,6 +133,18 @@ Organization getOrganization(int tenantId, String organizationId, String userId,
...
+
+        return new ArrayList<>();
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDao.java:145]: Why do we need a default implementation here?

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDaoImpl.java [hunk 6]
```diff
@@ -436,6 +437,26 @@ public void deleteOrganization(int tenantId, String organizationId) throws Organ
...
+            throws OrganizationManagementException {
+
+        JdbcTemplate jdbcTemplate = getNewTemplate();
+        try {
+            List<String> childOrganizationIds = jdbcTemplate.executeQuery(
+                    SQLConstants.FIND_ALL_CHILD_ORG_IDS,
+                    (resultSet, rowNumber) -> resultSet.getString(VIEW_ID_COLUMN),
+                    preparedStatement -> {
+                        int parameterIndex = 0;
+                        preparedStatement.setString(++parameterIndex, organizationId);
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/OrganizationMgtDaoImpl.java:451]: Use of Index parameter seems redundant.

File: org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/OrganizationUserRoleManagerImpl.java [hunk 7]
```diff
@@ -135,32 +136,34 @@ public void addOrganizationUserRoleMappings(String organizationId, UserRoleMappi
...
-            organizationList.add(organizationId);
-            organizationUserRoleMappings
-                    .addAll(populateOrganizationUserRoleMappings(organizationId, roleId, hybridRoleId, organizationId,
-                            usersGetPermissionsForSubOrgs));
-            while (!organizationList.isEmpty()) {
-                String currentOrgId = organizationList.remove();
-                List<String> children = cacheBackedOrganizationMgtDAO.getChildOrganizationIds(currentOrgId, null);
-                for (String childOrg : children) {
-                    organizationList.add(childOrg);
+        String useSpforInsert = System.getProperty(INSERT_ROLES_WITH_STORED_PROCEDURE);
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/OrganizationUserRoleManagerImpl.java:139]: Why do we read from a system property?

File: org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java [hunk 8]
```diff
@@ -23,23 +23,28 @@
...
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.organization.user.role.mgt.core.constant.SQLConstants;
 import org.wso2.carbon.identity.organization.user.role.mgt.core.exception.OrganizationUserRoleMgtException;
 import org.wso2.carbon.identity.organization.user.role.mgt.core.exception.OrganizationUserRoleMgtServerException;
-import org.wso2.carbon.identity.organization.user.role.mgt.core.model.OrganizationUserRoleMapping;
-import org.wso2.carbon.identity.organization.user.role.mgt.core.model.Role;
-import org.wso2.carbon.identity.organization.user.role.mgt.core.model.RoleAssignedLevel;
-import org.wso2.carbon.identity.organization.user.role.mgt.core.model.RoleAssignment;
-import org.wso2.carbon.identity.organization.user.role.mgt.core.model.RoleMember;
+import org.wso2.carbon.identity.organization.user.role.mgt.core.model.*;
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java:36]: wild card import

File: org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java [hunk 9]
```diff
@@ -119,6 +126,44 @@ public void addOrganizationUserRoleMappings(List<OrganizationUserRoleMapping> or
...
+
+                    callableStatement.addBatch();
+                }
+                //execute batch insert
+                callableStatement.executeBatch();
+                connection.commit();
+            } catch (SQLException e) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Error occurred while executing the batch insert: ", e);
+                }
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java:155]: Error will be logged anyway from the API layer.

File: org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java [hunk 10]
```diff
@@ -119,6 +126,44 @@ public void addOrganizationUserRoleMappings(List<OrganizationUserRoleMapping> or
...
+                connection.rollback();
+                throw handleServerException(ERROR_CODE_ORGANIZATION_USER_ROLE_MAPPINGS_ADD_ERROR, "", e);
```
Comment (nipunthathsara) [inline org.wso2.carbon.identity.organization.user.role.mgt.core/src/main/java/org/wso2/carbon/identity/organization/user/role/mgt/core/dao/OrganizationUserRoleMgtDAOImpl.java:157]: Pass null for data when not applicable.

File: org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java (hunk 2 above)
Comment (LahiruLS) [reply org.wso2.carbon.identity.organization.mgt.core/src/main/java/org/wso2/carbon/identity/organization/mgt/core/dao/CacheBackedOrganizationMgtDAO.java:276]: yes

**Pull request merged**